### PR TITLE
stm32f4 i2c bug fixes for 1.14 

### LINKED ISSUE #22751: STM32F407 I2C driver hangs
The STM32F407 I2C driver hangs from time to time where it waits on the semaphore in the driver.  When that happens the signal on the bus looks like this (SDA = yellow, SCL = green);

i2c_error

To work around it I added a timeout to the k_sem_take calls, but I have not found the real cause of the problem. On the scope is to see it must go wrong real early, because it generates a start-stop without any data. 

This is on the 1.14 branch, but I cherry-picked most commits (like the new timeout handling) from master. The target is a STM32F407 so it uses the V1 driver. Also not using IRQ's but polling causes the problem to happen less but it still happens.

### PR DESCRIPTION
**Pull request by lowlander**
With the STM32F4 I2C driver I ran into two problems;
1) the bug described in #22751 
2) EMC spikes cause the driver to hang so bad only a device reset can fix it.

Especially the resetting of the device is tricky, because it also resets all settings like bitrate. 

The fixes in these commits fix my problem, the hardware ran more than 300 hours nonstop with 10 I2C transactions per second. Without the fixes it doesn't last more than a few minutes before it hangs. 

The EMC problem only happened about every 30 hours, so a pain to debug, until I found out short circuiting SDA to GND will trigger it on command.

Fixes #22751.

### CODE REVIEW

**Comment by erwango** [comment]
^^ @ABOSTM

**Comment by carlescufi** [comment]
@lowlander has this been fixed in master?

File: drivers/i2c/i2c_ll_stm32_v1.c
Comment (lowlander) [reply drivers/i2c/i2c_ll_stm32_v1.c]: @ABOSTM busy is not supposed to be set at that point (and it will not go away on its own), what good would a timeout do? Or am I misunderstanding what you mean with "timeout" ?

**Comment by lowlander** [comment]
> @lowlander has this been fixed in master?

@carlescufi I don't know, the project I need it for still is on 1.14.  I can check after I get this patch done.

File: drivers/i2c/i2c_ll_stm32_v1.c
Comment (ABOSTM) [reply drivers/i2c/i2c_ll_stm32_v1.c]: @lowlander , There is some short transitional state where Busy could be set:
Suppose you write a first byte. Driver will write to Data register.
From Hardware point of view, data is transferred to shadow register before being transferred on I2C bus. During this phase, and until STOP, the bus is busy. But it is possible to prepare for next Byte transfer: the TXE bit is set, meaning it is possible to write the next Byte to Data register.
(hardware is able to hold 1st byte in shadow register and next byte in Data register).
Thus from software point of view, after writting the 1st byte, all the conditions are met to be able to write the next one. So you may have hardware busy state while calling `stm32_i2c_msg_write()` and busy state will go away on its own once physical transfer is finished (after STOP generation).

Comment (lowlander) [reply drivers/i2c/i2c_ll_stm32_v1.c]: @ABOSTM ok good point, any suggestions for the timeout? one byte at 100kHz  ~ 100uS ?

Comment (ABOSTM) [reply drivers/i2c/i2c_ll_stm32_v1.c]: yes, but adding some margin... and timeout should never occurs so why not 1ms

Comment (erwango) [reply drivers/i2c/i2c_ll_stm32_v1.c]: About timeout, there's a pending PR https://github.com/zephyrproject-rtos/zephyr/pull/21308
Can you check and see if it could be useful and/or superseded ?

**Comment by carlescufi** [comment]
> > @lowlander has this been fixed in master?
> 
> @carlescufi I don't know, the project I need it for still is on 1.14. I can check after I get this patch done.

CCB input: Before we can merge this PR we would need to know if this indeed requires a PR to master as well.

**Comment by nashif** [comment]
@erwango  does this issue exist also in master?

**Comment by erwango** [comment]
@carlescufi, @nashif, This issue is not fixed in master.
Some of the changes are there (timeout in first commit), but not the core of the fix.
Also, it could be noted that files have diverged, so fix has to be reworked. 
Btw, on master, I think implementation of https://github.com/zephyrproject-rtos/zephyr/issues/23441 would be a solution to reported issue (using parts of this change).

@lowlander, one question, do you know if error is anyhow specific to v1?

**Comment by henrikbrixandersen** [comment]
I encountered a similar issue on NXP Kinetis a while back. The NXP LPSPI controllers supports an idle timeout, which I implemented support for here: #20626

**Comment by MaureenHelm** [comment]
@erwango will you please update this issue? will it get fixed in master?

**Comment by lowlander** [comment]
Hey All, sorry that I didn't find more time to move this PR forward. Our system now has been in use for a while and we had no complains about I2c anymore. But I must say I have build in a "hard-reset" in our system, where I use GPIO to reset the I2C bus. Building that into the 1.14 driver isn't that easy because the driver doesn't have easy access to the GPIO's. 

I have not been able to test master since then I would first have to port master to our hardware :-/ 

So not really sure how to move forward with this, would be good if there are some other 1.14/I2C users that could test it if it at least doesn't break anything.

**Comment by lowlander** [comment]
@MaureenHelm I updated this fix for master in #27291 can you give it a look ?

**Comment by MaureenHelm** [comment]
> @MaureenHelm I updated this fix for master in #27291 can you give it a look ?

Thank you!

**Review by ydamigos — APPROVED** [review]

**Comment by pabigot** [comment]
> Hey All, sorry that I didn't find more time to move this PR forward. Our system now has been in use for a while and we had no complains about I2c anymore. But I must say I have build in a "hard-reset" in our system, where I use GPIO to reset the I2C bus. Building that into the 1.14 driver isn't that easy because the driver doesn't have easy access to the GPIO's.

@lowlander Based on this comment, and that master has already merged your fix, do you need this in 1.14 anymore or can we just close it?

There're enough changes between this and the original and master that I'm not really comfortable approving without testing, and I don't have hardware for this variant.  So if we don't need it backported that'd be great.

**Comment by lowlander** [comment]
> @lowlander Based on this comment, and that master has already merged your fix, do you need this in 1.14 anymore or can we just close it?

@pabigot I moved my projects to master, so personally I will not need it on 1.14 anymore, but it might bite others.

**Pull request closed without merging**
